src/coreclr/jit/importer.cpp

```diff
@@ -15665,7 +15665,13 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                                                            constraintCall ? &constrainedResolvedToken : nullptr,
                                                            returnFalseIfInvalid);
 
-                            if (passedConstraintCheck)
+                            // Avoid setting compHasBackwardsJump = true via tail call stress if the method cannot have
+                            // patchpoints.
+                            //
+                            const bool mayHavePatchpoints = opts.jitFlags->IsSet(JitFlags::JIT_FLAG_TIER0) &&
+                                                            (JitConfig.TC_OnStackReplacement() > 0) &&
+                                                            compCanHavePatchpoints();
+                            if (passedConstraintCheck && (mayHavePatchpoints || compHasBackwardJump))
```

Comment:
The fix looks fine to me but I am a bit confused we are marking a backward jump in the first place. We should not do the loop optimization in tier-0 (and, since #41059, never under tailcall stress).

I can clean this up in the future if you don't want to do this as part of this PR.

Reply:
But clearly you had to add special support for OSR due to the tailcall-to-loop optimization, so maybe I am missing something?

Reply:
No, you're on to something.

We are probably too aggressive in marking explicit tail calls as backwards branches in Tier0&mdash;and more importantly in this case, changing the value of `compHasBackwardJump`. We do this regardless of optimization flags. If we are never going to do the optimization, then we don't need to change `compHasBackwardJump` either.

Another possible fix would be that when tail call stress changes an implicit tail call to an explicit one, it should also set `compTailPrefixSeen`. But it's not as clean of a fix; we might already have added some patchpoints.